### ISSUE TITLE
Create key/value map in WebSocket

### ISSUE THREAD

**Issue by rjbergTU** [TooTallNate/Java-WebSocket#485]
It would be nice if each WebSocket had a map for key/value pairs. This way you can keep track of different clients in a simple way. I have seen other server classes in other projects using this kind of functionality.

**Comment by marci4** [comment]
Hello @rjbergTU,

first of all thanks a lot for your input :) 

I have the following "problems" with your idea:
- Does everybody need this? (Every websocket server keeping a map containing websocket clients keeping a map)
- What are the best key/value classes? (Probably resulting into <Object,Object> so everybody has to cast their results)

Greetings
marci4

**Comment by Irahe** [comment]
Well,

On my project i use this kind of mapping for authentication and identification proposes. 
I have created two separated chasses and two Synchronized lists on the Server class, so i can keep track of this.
I believe this is not the role of the lib, but a specificity of each implementation.

Greetings. :*

**Comment by rjbergTU** [comment]
Hello @marci4 

Regarding your feedback, I have the following things:

1.a No, not everybody needs this. But I do believe it is very common to be able to distinguish between clients and assign them variables.
1.b I actually was more talking about a map in the WebSocket implementations then one in the server or client. This makes it easier then a map which maps WebSockets to other maps.

2. I think it is no problem if keys have to be Strings. It is also more useful with map implementations that need instances of Comparable, like TreeMap. I would recommend using Objects for values though.

3. Maybe, but it is a bit of a hassle to modify WebsocketImpl, or to assign a map in each server mapping WebSocket to other maps, which need their own keys. It other words I think it is easier to be able to something like webSocket.get(key) in the server instead of socketMap.get(webSocket).get(key).

**Comment by marci4** [comment]
_Concept and implementation for future release!_

**Comment by lifubang** [comment]
I think if the maintainer provide a clientid field in org.java_websocket.WebSocketImpl class, the map can be maintained by ourselves. (#554)

But I also think that because the class org.java_websocket.server.WebSocketServer provide a private field Collection<WebSocket> connections, if this field changes to Dictionary<String, WebSocket> connections, It'll be more Convenient for us. But it will need more code changes. So I think it's not a good idea.

So I think you can add a clientid field in org.java_websocket.WebSocketImpl class, the problem will be resolved with a lowest cost.

**Comment by 8databit** [comment]
I agree with lifubang, add an getter/setter in the class mentioned for just an Object
`private Object attachment = null;`

This is the best way sinse we dont have to find the address of the object inside an iterator of some sort. 
Maybe add it to the websocket interface too, would be good sinse it's somewhat a standard thing to add custom information/references to a socket.

I highly disagree with adding a map. It isn't necessary sinse this can be easily implemented with help of the onOpen() method of the server.

**Comment by marci4** [comment]
Hello Joyie,

Everything mentioned here (Map or Object) can be implemented on a application level as well!

The question is what is the best approach for most of the users of this project!

Greetings
marci4

**Comment by rjbergTU** [comment]
Now that I checked all this, I agree that a Map per websocket might be overkill for many applications. However, I do believe some way to distinguish between clients in the server is needed. A clientid field should suffice for this.

However, I do have a small warning: If multiple clients are handled in different threads, it might be handy to have a map for each client, as my initial proposal. The reason is that multithreading might break a standard(not thread safe) Map of client id's to other objects. Using clientid's is more lightweight then a TreeMap<String, Object> for each client, but also more risky, as a coder might use a non thread safe map.

Overall, both solutions have their merits to me. The map solution might be safer, but overkill in some situations. The client id is more lightweight, but also harder to use in some cases, or have more work to it.

**Comment by promatik** [comment]
Hi everyone!
I just created a pull request for this feature #627.

After reading all your comments I ran some tests between having an `HashMap` or creating a getter and a setter to an object, and clearly `HashMap` was a lot worst, even with fewer connections.

My implementation is pretty simple to use, here is an example on how to set the attachment on `onOpen` and access it on `onMessage`:

```java
public void onOpen(WebSocket conn, ClientHandshake handshake) {
    User user = new User(conn);
    conn.setAttachment(user);
}

public void onMessage(WebSocket conn, String message) {
    User user = conn.getAttachment();
    // ...
}
```

I'm using it just now, and it's working great.

**Comment by marci4** [comment]
This issue is implemented with #627.

Thanks a lot to @promatik for his contribution!

Greetings
marci4

**Comment by rjbergTU** [comment]
I am not surprised a HashMap per connection is bad, since a HashMap needs a lot of memory. (One reason I do not like them, except in cases where the number and total size of the Map can be predetermined.) But amap of any kind(like a TreeMap) would be overkill anyway. Your current solution is a lot more flexible. 

In short, thank you.
